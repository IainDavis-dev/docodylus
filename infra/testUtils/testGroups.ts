import { describe } from 'vitest';

// types set-up
const TestGroups = [
  'ALL',
  'UNIT',
  'INTEGRATION',
  'POLICY',
  'TYPE',
] as const;
type TestGroup = typeof TestGroups[number];

// run all tests -- default behavior
const ALL = TestGroups[0];

// run unit tests
const UNIT = TestGroups[1];

// run tests that validate interoperability of modules
const INTEGRATION = TestGroups[2];

// run tests that enforce conventions
const POLICY = TestGroups[3];

// run tests that validate typescript types.
const TYPE = TestGroups[4];

// boilerplate functions
function isTestGroup(maybeTestGroup: string): maybeTestGroup is TestGroup {
  return TestGroups.some((tg) => tg === maybeTestGroup);
}

const testGroups: TestGroup[] = process.env.TEST_GROUP?.split(',').filter(isTestGroup) || [ALL];
if (testGroups.length === 0) testGroups.push(ALL);

function checkGroups(groups: TestGroup | TestGroup[]): boolean {
  const resolvedGroups = Array.isArray(groups) ? groups : [groups];
  return [...resolvedGroups, ALL].some((g) => testGroups.includes(g));
}

// helper functions (describe, but with in-advance currying to set the test group)
/**
 * Supplies the necessary boilerplate to ensure a block of tests runs
 * only when the UNIT test group (or none) is selected
 */
export const describeUnitTest = describe.runIf(checkGroups(UNIT));

/**
 * Supplies the necessary boilerplate to ensure a block of tests runs
 * only when the INTEGRATION test group (or none) is selected
 */
export const describeIntegrationTest = describe.runIf(checkGroups(INTEGRATION));

/**
 * Supplies the necessary boilerplate to ensure a block of tests runs
 * only when the POLICY test group (or none) is selected
 */
export const describePolicyTest = describe.runIf(checkGroups(POLICY));

/**
 * Supplies the necessary boilerplate to ensure a block of tests runs
 * only when the TYPE test group (or none) is selected
 */
export const describeTypeTest = describe.runIf(checkGroups(TYPE));
