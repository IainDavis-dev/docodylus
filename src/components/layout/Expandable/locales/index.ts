import { BASE_NAMESPACE } from "@i18n/consts";
import { FQTxlns } from "@i18n/types";
import { createNamespacePrepender } from "@i18n/utils/createNamespacePrepender";

export const ExpandableNamespace = `${BASE_NAMESPACE}.layout.expandable` as const;
export const prependNamespace = createNamespacePrepender(ExpandableNamespace);

export type ExpandableLocalizedStrings = FQTxlns<typeof ExpandableNamespace, {
    expandPrompt: string
    collapsePrompt: string
}>

export const namespacer = (raw: string) => `${ExpandableNamespace}.${raw}`;
