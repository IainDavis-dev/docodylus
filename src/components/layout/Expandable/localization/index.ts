import { BASE_NAMESPACE } from "@i18n/consts";
import { getRawLoaders } from "./getRawLoaders";
import { createNamespacePrepender } from "@shared-utils/createNamespacePrepender";
import { Namespaced } from "@shared-utils/types";
import { toLocalizationFileLoaderMap } from "@i18n/utils/localeNegotiation";

export const ExpandableNamespace = `${BASE_NAMESPACE}.layout.expandable` as const;
export const prependNamespace = createNamespacePrepender(ExpandableNamespace);


export type ExpandableLocalizedStrings = Namespaced<typeof ExpandableNamespace, {
    expandPrompt: string
    collapsePrompt: string
}>

/**
 * Takes the map of {relative-URL} -> {translation-file-lazy-loader},
 * and converts the keys so that they are absolute paths which can be
 * used as cache keys
 * 
 * @returns a map with keys that are aboslute URLs and values which are functions to dynamically load the files
 */
export function getLocalizedStringLoaders() {
    const fileLoaders = getRawLoaders();
    const normalizedFileLoaders = Object.entries(fileLoaders).reduce(
        // URL generation here depends on the location of this file, so we
        // cannot readily externalize this
        (obj, [key, val]) => ({...obj, [new URL(key, import.meta.url).href]: val}),
        {}
    );
    return toLocalizationFileLoaderMap<ExpandableLocalizedStrings>(normalizedFileLoaders);
}

// adds localized-string keys for this component to the master list so
// it can be included in user overrides.
declare module "@i18n/types" {
    export interface DocodylusLocalizedStrings extends ExpandableLocalizedStrings {}
}