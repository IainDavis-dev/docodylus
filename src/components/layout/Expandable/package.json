{
    "name": "@docodylus/Expandable",
    "version": "0.0.1-prerelease.3",
    "description": "A component whose children are contained in a region that is concealable or revealable with the use of a toggle",
    "type": "module",
    "exports": {
        "import": "./dist/Expandable.mjs",
        "require": "./dist/Expandable.cjs"
    },
    "main": "./dist/Expandable.cjs",
    "module": "./distExpandable.mjs",
    "types": "./dist/Expandable.d.ts",
    "sideEffects": false,
    "peerDependencies": {
        "react": "^19.0.0"
    },
    "devDependencies": {
        "react": "^19.0.0",
        "typescript": "^5.7.2"
    }
}
