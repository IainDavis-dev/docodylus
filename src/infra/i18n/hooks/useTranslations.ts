import { DEFAULT_TRANSLATIONS, LOADING_KEY } from "@i18n/consts";
import { I18nContext } from "@i18n/context/I18nContext";
import LocaleAwarePolyglot from "@i18n/LocaleAwarePolyglot/LocaleAwarePolyglot";
import { asValidLocale, ProvidedLocales, Txlns } from "@i18n/types";
import { PolyglotOptions } from "node-polyglot";
import { useContext, useEffect, useMemo, useState } from "react";

// default instance serves DEFAULT_LOCALE in the absence of an explicit Provider
let fallbackPolyglot: LocaleAwarePolyglot;

type DefaultTranslationKey = keyof typeof DEFAULT_TRANSLATIONS['en'];
type LoadingState = 'not-loaded' | 'loading' | 'success' | 'error';
type TWrapper<T> = (key: T | DefaultTranslationKey, options?: PolyglotOptions) => string;

/**
 * A React hook that ensures translations are loaded into
 * {@link LocaleAwarePolyglot} before translations are invoked.
 * 
 * @template T - Type representing the translation keys. Used only to provide
 * code-completion hints for the set of translations relevant to the compoenent
 *
 * @param {URL} translationsSrc - The URL of the translation file to load.
 * @returns {UseTranslationsResponse<T>} An object containing:
 * - `t`: A function to retrieve translated strings by key.
 * - `isLoading`: A boolean indicating whether translations are currently being loaded.
 */
const useTranslations = <T extends Txlns = never>(providedLocales: ProvidedLocales<T>): TWrapper<keyof T>  => {
    const context = useContext(I18nContext) ?? { i18n: fallbackPolyglot ??= new LocaleAwarePolyglot()};

    const { i18n = fallbackPolyglot } = context;
    const  locale = i18n.getLocale();

    const [loadingStates, setLoadingStates] = useState<Record<string, LoadingState>>(() => Object.create(null))

    const negotiatedLocales = Object.entries(providedLocales).reduce<ProvidedLocales<T>>(
        (negotiated, [key, val]) => locale.startsWith(key) ? {...negotiated, [key]: val} : negotiated,
        {}
    )
    
    useEffect(() => {
        Object.entries(negotiatedLocales)
            .filter(([_, { cacheKey }]) => {
                const currentState: LoadingState = loadingStates[cacheKey] ?? 'not-loaded';
                return !['loading', 'success', 'error'].includes(currentState);
            // map here, rather than forEach, to ensure loaders run synchronously
            }).map(([locale, { cacheKey, loader }]) => {
                const wrappedLoader = async () => {
                    setLoadingStates((prev) => ({ ...prev, [cacheKey]: 'loading'}));
                    const validLocale = asValidLocale(locale)
                    if(!validLocale) return

                    try {
                        const localizedStrings = (await loader()).default as unknown as T;
                        setLoadingStates((prev) => ({ ...prev, [cacheKey]: 'success'}));
                        i18n.extend(validLocale, localizedStrings);
                    } catch (error) {
                        console.error(`Failed to load translations from ${cacheKey}:`, error);
                        setLoadingStates((prev) => ({ ...prev, [cacheKey]: 'error' }))
                    }
                }
                wrappedLoader();
            })
        }, [locale, providedLocales]);

    const usePlaceholderText = Object.entries(negotiatedLocales).some(([locale, { cacheKey }]) => {
        const currentLoadingState = loadingStates[cacheKey];
        return ['not-loaded', 'loading'].includes(currentLoadingState);
    })

    const tWrapper: TWrapper<keyof T> = useMemo(() =>
        (key, options) => usePlaceholderText ? i18n.t(LOADING_KEY) : i18n.t(key as string, options),
        [usePlaceholderText] // no need to continually re-create this function
    );

    return tWrapper;
}

export default useTranslations;
