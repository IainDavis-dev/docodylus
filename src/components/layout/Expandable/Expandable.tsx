import React, { KeyboardEventHandler, MouseEventHandler, PropsWithChildren,  useEffect,  useId,  useRef,  useState } from 'react';
import styles from './Expandable.module.css';
import useTranslations from '@i18n/hooks/useTranslations';

import { ExpandableLocalizedStrings as ExpandableTxlns, ExpandableNamespace as ns } from './localization';
import { ProvidedLocales } from '@i18n/types';
import { getProvidedLocales } from '@i18n/utils/localeNegotiation';

export type ExpandablePropsType = {
    startExpanded?: boolean;
    expandPrompt?: string;
    collapsePrompt?: string;
}

const providedLocales: ProvidedLocales<ExpandableTxlns> = getProvidedLocales(import.meta.glob<Partial<ExpandableTxlns>>('./localization/*.txlns.ts'));

const Expandable: React.FC<PropsWithChildren<ExpandablePropsType>> = ({
    startExpanded = false,
    expandPrompt,
    collapsePrompt,
    children
}) => {
    const t = useTranslations<ExpandableTxlns>(providedLocales);

    const idDiscriminator = useId();

    const contentRef = useRef<HTMLDivElement>(null);
    const buttonRef = useRef<HTMLButtonElement>(null);

    const effectiveExpandPrompt =  expandPrompt || t(`${ns}.expandPrompt`);
    const effectiveCollapsePrompt = collapsePrompt || t(`${ns}.collapsePrompt`);

    const [isExpanded, setExpanded] = useState(startExpanded);

    useEffect(() => {
        setExpanded(() => startExpanded);
    }, [startExpanded])

    useEffect(() => {
        if(isExpanded) { contentRef.current?.focus(); }
        else { buttonRef.current?.focus(); }
    })

    const toggleExpanded = () => setExpanded(prevExpanded => !prevExpanded);

    const handleClick: MouseEventHandler = (e) => {
        toggleExpanded();
    }

    const handleKeyDown: KeyboardEventHandler = (e) => {
        if(['Enter', ' '].includes(e.key)) {
            e.preventDefault();
            toggleExpanded();
        }
    }

    return (
        <div className={styles.container}>
            <div
                ref={contentRef}
                tabIndex={isExpanded ? 0 : 1}
                data-testid={`expandable-section-${idDiscriminator}`}
                id={`expandable-section-${idDiscriminator}`}
                className={ styles.expandableBlock }
                hidden={!isExpanded}
            >
                {children}
            </div>
            <button
                ref={buttonRef}
                tabIndex={0}
                className={styles.toggle}
                onClick={handleClick}
                onKeyDown={handleKeyDown}
                aria-expanded={isExpanded}
                aria-controls={`expandable-section-toggle-${idDiscriminator}`}
            >
                { isExpanded ? effectiveCollapsePrompt : effectiveExpandPrompt }
            </button>
        </div>
    )
}

export default Expandable;