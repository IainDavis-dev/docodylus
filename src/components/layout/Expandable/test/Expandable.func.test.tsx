import { describeUnitTest } from '@test-utils/testGroups';
import { mockUseTranslations, MOCK_DEFAULT_EXPAND_PROMPT, MOCK_DEFAULT_COLLAPSE_PROMPT } from './__mocks__/useTranslations';

vi.mock("import.meta.glob", () => Promise.resolve({})); // no-op
vi.mock("@i18n/hooks/useTranslations", () => ({ default: mockUseTranslations }))

import { fireEvent, render, screen } from '@testing-library/react';
import Expandable from '../Expandable';

describeUnitTest('Core functionality tests', () => {
    describe('Initial state', () => {
        it('should render collapsed by default and expand when clicked', () => {
            render(<Expandable />);
            expect(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT)).toBeInTheDocument();
        });

        it('should render collapsed when startExpanded is explicitly false', () => {
            render(<Expandable startExpanded={false}/>)
            expect(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT)).toBeInTheDocument();
        })

        it('should render expanded when startExpanded is explicitly true', () => {
            render(<Expandable startExpanded={true} />);
            expect(screen.getByText(MOCK_DEFAULT_COLLAPSE_PROMPT)).toBeInTheDocument();
        });
    })

    describe('State change', () => {

        it('should collapse when button is clicked from an expanded state', () => {
            render(<Expandable startExpanded={true}/>);
            const content = screen.getByTestId(/^expandable-section/);
            expect(content).toBeVisible()

            fireEvent.click(screen.getByText(MOCK_DEFAULT_COLLAPSE_PROMPT));

            expect(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT)).toBeInTheDocument();
            expect(content).not.toBeVisible()
        });

        it('should expand when button is clicked from a collapsed state', () => {
            render(<Expandable startExpanded={false}/>);
            const content = screen.getByTestId(/^expandable-section/);
            expect(content).not.toBeVisible()

            fireEvent.click(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT));

            expect(screen.getByText(MOCK_DEFAULT_COLLAPSE_PROMPT)).toBeInTheDocument();
            expect(content).toBeVisible()
        })

    })

    describe("Customization props", () => {
        it('should handle empty prompts gracefully', () => {
            render(<Expandable startExpanded={false} expandPrompt="" collapsePrompt="" />);
            fireEvent.click(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT));
            expect(screen.getByText(MOCK_DEFAULT_COLLAPSE_PROMPT)).toBeInTheDocument();
        });

        it('should handle null prompts gracefully', () => {
            render(<Expandable startExpanded={false} expandPrompt={null as any} collapsePrompt={null as any} />);
            fireEvent.click(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT));
            expect(screen.getByText(MOCK_DEFAULT_COLLAPSE_PROMPT)).toBeInTheDocument();
        });

        it('should handle explicitlly undefined prompts gracefully', () => {
            render(<Expandable startExpanded={false} expandPrompt={undefined} collapsePrompt={undefined}/>);
            fireEvent.click(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT));
            expect(screen.getByText(MOCK_DEFAULT_COLLAPSE_PROMPT)).toBeInTheDocument();
        });

        it('should use custom prompts when provided', () => {
            const MOCK_CUSTOM_EXPAND_PROMPT = "CUSTOM_EXPAND_PROMPT";
            const MOCK_CUSTOM_COLLAPSE_PROMPT = "CUSTOM_COLLAPSE_PROMPT";

            render(<Expandable startExpanded={false} expandPrompt={MOCK_CUSTOM_EXPAND_PROMPT} collapsePrompt={MOCK_CUSTOM_COLLAPSE_PROMPT}/>);
            expect(screen.getByText(MOCK_CUSTOM_EXPAND_PROMPT)).toBeInTheDocument();
            expect(screen.queryByText(MOCK_DEFAULT_EXPAND_PROMPT)).not.toBeInTheDocument();

            fireEvent.click(screen.getByText(MOCK_CUSTOM_EXPAND_PROMPT));
            expect(screen.getByText(MOCK_CUSTOM_COLLAPSE_PROMPT)).toBeInTheDocument();
            expect(screen.queryByText(MOCK_DEFAULT_COLLAPSE_PROMPT)).not.toBeInTheDocument();
        })

        it('should update when startExpanded prop is changed dynamically', () => {
            const { rerender } = render(<Expandable startExpanded={false} />);
            expect(screen.queryByText(MOCK_DEFAULT_EXPAND_PROMPT)).toBeInTheDocument();
            expect(screen.queryByText(MOCK_DEFAULT_COLLAPSE_PROMPT)).not.toBeInTheDocument

            rerender(<Expandable startExpanded={true} />);
            expect(screen.getByText(MOCK_DEFAULT_COLLAPSE_PROMPT)).toBeInTheDocument();
            expect(screen.queryByText(MOCK_DEFAULT_EXPAND_PROMPT)).not.toBeInTheDocument();
        });
    });

    describe("Snapshots/Visual-Regression", () => {
        it('should maintain visual consistency between expanded and collapsed states', () => {
            const { asFragment, rerender } = render(<Expandable startExpanded={false} />);
            expect(asFragment()).toMatchSnapshot();

            rerender(<Expandable startExpanded={true} />);
            expect(asFragment()).toMatchSnapshot();
        });
    })

    // Test large content performance
    it('should perform well with large content when expanding/collapsing', () => {
        const largeContent = Array(1000).fill('Content').join(' ');
        render(<Expandable>{largeContent}</Expandable>);
        fireEvent.click(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT));
        expect(screen.getByText(MOCK_DEFAULT_COLLAPSE_PROMPT)).toBeInTheDocument();
    });

    // Check responsiveness/media queries behavior
    it('should respond correctly to different screen sizes', () => {
        global.innerWidth = 500;
        render(<Expandable />);
        expect(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT)).toBeInTheDocument();

        global.innerWidth = 1024;
        fireEvent.resize(window);
        expect(screen.getByText(MOCK_DEFAULT_EXPAND_PROMPT)).toBeInTheDocument();
    });
});
    