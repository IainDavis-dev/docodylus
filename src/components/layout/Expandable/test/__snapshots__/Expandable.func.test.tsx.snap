// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Core functionality tests > Snapshots/Visual-Regression > should maintain visual consistency between expanded and collapsed states 1`] = `
<DocumentFragment>
  <div
    class="_container_6cd2f2"
  >
    <div
      class="_expandableBlock_6cd2f2"
      data-testid="expandable-section-:ra:"
      hidden=""
      id="expandable-section-:ra:"
      tabindex="1"
    />
    <button
      aria-controls="expandable-section-toggle-:ra:"
      aria-expanded="false"
      class="_toggle_6cd2f2"
      tabindex="0"
    >
      MOCK_COLLAPSE_PROMPT
    </button>
  </div>
</DocumentFragment>
`;

exports[`Core functionality tests > Snapshots/Visual-Regression > should maintain visual consistency between expanded and collapsed states 2`] = `
<DocumentFragment>
  <div
    class="_container_6cd2f2"
  >
    <div
      class="_expandableBlock_6cd2f2"
      data-testid="expandable-section-:ra:"
      id="expandable-section-:ra:"
      tabindex="0"
    />
    <button
      aria-controls="expandable-section-toggle-:ra:"
      aria-expanded="true"
      class="_toggle_6cd2f2"
      tabindex="0"
    >
      MOCK_EXPAND_PROMPT
    </button>
  </div>
</DocumentFragment>
`;
